src/vs/workbench/contrib/positronConnections/browser/positronConnections.tsx

```diff
@@ -0,0 +1,276 @@
+import { DisposableStore } from 'vs/base/common/lifecycle';
+import { useStateRef } from 'vs/base/browser/ui/react/useStateRef';
+import * as DOM from 'vs/base/browser/dom';
+import { IPositronConnectionEntry } from 'vs/workbench/services/positronConnections/browser/positronConnectionsCache';
+
+export interface PositronConnectionsProps {
+	readonly commandService: ICommandService;
+	readonly configurationService: IConfigurationService;
+	readonly contextKeyService: IContextKeyService;
+	readonly contextMenuService: IContextMenuService;
+	readonly hoverService: IHoverService;
+	readonly keybindingService: IKeybindingService;
+	readonly connectionsService: IPositronConnectionsService;
+	readonly reactComponentContainer: IReactComponentContainer;
+}
+
+export const PositronConnections = (props: React.PropsWithChildren<PositronConnectionsProps>) => {
+
+	// This allows us to introspect the size of the component. Which then allows
+	// us to efficiently only render items that are in view.
+	const [_, setWidth] = React.useState(props.reactComponentContainer.width);
+	const [height, setHeight] = React.useState(props.reactComponentContainer.height);
+
+	useEffect(() => {
+		const disposableStore = new DisposableStore();
+		disposableStore.add(props.reactComponentContainer.onSizeChanged(size => {
+			setWidth(size.width);
+			setHeight(size.height);
+		}));
+		return () => disposableStore.dispose();
+	}, []);
+
+	// We're required to save the scroll state because browsers will automatically
+	// scrollTop when an object becomes visible again.
+	const [, setScrollState, scrollStateRef] = useStateRef<number[] | undefined>(undefined);
+	const innerRef = useRef<HTMLElement>(undefined!);
+	useEffect(() => {
+		const disposableStore = new DisposableStore();
+		disposableStore.add(props.reactComponentContainer.onSaveScrollPosition(() => {
+			if (innerRef.current) {
+				setScrollState(DOM.saveParentsScrollTop(innerRef.current));
+			}
+		}));
+		disposableStore.add(props.reactComponentContainer.onRestoreScrollPosition(() => {
+			if (scrollStateRef.current) {
+				if (innerRef.current) {
+					DOM.restoreParentsScrollTop(innerRef.current, scrollStateRef.current);
+				}
+				setScrollState(undefined);
+			}
+		}));
+		return () => disposableStore.dispose();
+	}, []);
+
+	const [items, setItems] = useState<IPositronConnectionEntry[]>(props.connectionsService.getConnectionEntries);
+	useEffect(() => {
+		const disposableStore = new DisposableStore();
+		disposableStore.add(props.connectionsService.onDidChangeEntries((entries) => {
+			setItems(entries);
+		}));
+		// First entries refresh - on component mount.
+		props.connectionsService.refreshConnectionEntries();
+		return () => disposableStore.dispose();
+	}, []);
+
+	const [selectedId, setSelectedId] = useState<string>();
+
+	const ItemEntry = (props: ItemEntryProps) => {
+		const itemProps = items[props.index];
+
+		return (
+			<PositronConnectionsItem
```

Comment:
Perhaps:

```
		return (
			<PositronConnectionsItem
				{...itemProps}
				selected={itemProps.id === selectedId}
				onSelectedHandler={() => setSelectedId(itemProps.id)}
				style={props.style}>
			</PositronConnectionsItem>
		);
```

Just a thought.

Reply:
I tried this and it doesn't work. I have no idea why :S
I just changed to https://github.com/posit-dev/positron/pull/4967/commits/c34d6c7118ef540d36288da17cc666802221ae34 which looks slightly better, I think!